Note this notebook is for the Question no 2 and 3

# Setup

In [1]:
!pip install tensorflow==2.0.0-rc0

from sklearn.model_selection import train_test_split
from sklearn.datasets import load_boston
import tensorflow as tf
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
sns_colors = sns.color_palette('colorblind')

tf.random.set_seed(1)

     |████████████████████████████████| 86.3MB 108kB/s 
     |████████████████████████████████| 51kB 6.6MB/s 
     |████████████████████████████████| 501kB 46.0MB/s 
     |████████████████████████████████| 4.3MB 43.6MB/s 
  Found existing installation: tensorflow 2.3.0
    Uninstalling tensorflow-2.3.0:
      Successfully uninstalled tensorflow-2.3.0


# Importing the dataset

In [2]:

boston_dataset = load_boston()

data_X = pd.DataFrame(boston_dataset.data, columns=boston_dataset.feature_names)
data_Y = pd.DataFrame(boston_dataset.target, columns=["target"])
data = pd.concat([data_X, data_Y], axis=1)

In [3]:
train, test = train_test_split(data, test_size=0.2, random_state=1)
print(len(train), "train examples")
print(len(test), "test examples")


404 train examples
102 test examples


In [4]:
def df_to_dataset(dataframe, shuffle_and_repeat=True, batch_size=32):
  dataframe = dataframe.copy()
  labels = dataframe.pop('target')
  ds = tf.data.Dataset.from_tensor_slices((dict(dataframe), labels))
  if shuffle_and_repeat:
    ds = ds.shuffle(buffer_size=len(dataframe))
    ds = ds.repeat()
    
  ds = ds.batch(batch_size)
  
  return ds

In [5]:
test.head()

CRIM    ZN  INDUS  CHAS    NOX  ...    TAX  PTRATIO       B  LSTAT  target
307  0.04932  33.0   2.18   0.0  0.472  ...  222.0     18.4  396.90   7.53    28.2
343  0.02543  55.0   3.78   0.0  0.484  ...  370.0     17.6  396.90   7.18    23.9
47   0.22927   0.0   6.91   0.0  0.448  ...  233.0     17.9  392.74  18.80    16.6
67   0.05789  12.5   6.07   0.0  0.409  ...  345.0     18.9  396.21   8.10    22.0
362  3.67822   0.0  18.10   0.0  0.770  ...  666.0     20.2  380.79  10.19    20.8

[5 rows x 14 columns]

In [6]:
train_y=train.pop('target')
test_y=test.pop('target')

In [8]:
def input_fn(features,labels,training=True,batch_size=256):
  dataset=tf.data.Dataset.from_tensor_slices((dict(features),labels))
  if training:
    dataset=dataset.shuffle(1000).repeat()
  return dataset.batch(batch_size)

**Defining Feature Columns**

In [9]:
feature_columns = []
for key in data_X.keys():
  if key == 'RAD':
    continue
  feature_columns.append(tf.feature_column.numeric_column(key))

feature_columns

[NumericColumn(key='CRIM', shape=(1,), default_value=None, dtype=tf.float32, normalizer_fn=None),
 NumericColumn(key='ZN', shape=(1,), default_value=None, dtype=tf.float32, normalizer_fn=None),
 NumericColumn(key='INDUS', shape=(1,), default_value=None, dtype=tf.float32, normalizer_fn=None),
 NumericColumn(key='CHAS', shape=(1,), default_value=None, dtype=tf.float32, normalizer_fn=None),
 NumericColumn(key='NOX', shape=(1,), default_value=None, dtype=tf.float32, normalizer_fn=None),
 NumericColumn(key='RM', shape=(1,), default_value=None, dtype=tf.float32, normalizer_fn=None),
 NumericColumn(key='AGE', shape=(1,), default_value=None, dtype=tf.float32, normalizer_fn=None),
 NumericColumn(key='DIS', shape=(1,), default_value=None, dtype=tf.float32, normalizer_fn=None),
 NumericColumn(key='TAX', shape=(1,), default_value=None, dtype=tf.float32, normalizer_fn=None),
 NumericColumn(key='PTRATIO', shape=(1,), default_value=None, dtype=tf.float32, normalizer_fn=None),
 NumericColumn(key='B', 

# Building the model

In [10]:
config = tf.estimator.RunConfig(tf_random_seed=1)

model=tf.estimator.DNNRegressor(
    feature_columns=feature_columns,
    hidden_units=[15],
    config=config
)

INFO:tensorflow:Using config: {'_model_dir': '/tmp/tmp87tla0d2', '_tf_random_seed': 1, '_save_summary_steps': 100, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_session_config': allow_soft_placement: true
graph_options {
  rewrite_options {
    meta_optimizer_iterations: ONE
  }
}
, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_train_distribute': None, '_device_fn': None, '_protocol': None, '_eval_distribute': None, '_experimental_distribute': None, '_experimental_max_worker_delay_secs': None, '_service': None, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7f2df2bf5160>, '_task_type': 'worker', '_task_id': 0, '_global_id_in_cluster': 0, '_master': '', '_evaluation_master': '', '_is_chief': True, '_num_ps_replicas': 0, '_num_worker_replicas': 1}


**Train the model**

In [11]:
model.train(
input_fn=lambda:input_fn(train,train_y,training=True),
    steps=5000
)

Instructions for updating:
If using Keras pass *_constraint arguments to layers.
Instructions for updating:
Use Variable.read_value. Variables in 2.X are initialized automatically both in eager and graph (inside tf.defun) contexts.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
INFO:tensorflow:Calling model_fn.

If you intended to run this layer in float32, you can safely ignore this warning. If in doubt, this warning is likely only an issue if you are porting a TensorFlow 1.X model to TensorFlow 2.

To change all layers to have dtype float64 by default, call `tf.keras.backend.set_floatx('float64')`. To change just this layer, pass dtype='float64' to the layer constructor. If you are the author of this layer, you can disable autocasting by passing autocast=False to the base Layer constructor.

Instructions for updating:
Use `tf.cast` instead.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it

# Evaluate on test data

In [12]:
model.evaluate(
input_fn=lambda:input_fn(test,test_y,training=False))

INFO:tensorflow:Calling model_fn.

If you intended to run this layer in float32, you can safely ignore this warning. If in doubt, this warning is likely only an issue if you are porting a TensorFlow 1.X model to TensorFlow 2.

To change all layers to have dtype float64 by default, call `tf.keras.backend.set_floatx('float64')`. To change just this layer, pass dtype='float64' to the layer constructor. If you are the author of this layer, you can disable autocasting by passing autocast=False to the base Layer constructor.

INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2020-10-27T03:39:03Z
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from /tmp/tmp87tla0d2/model.ckpt-5000
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Finished evaluation at 2020-10-27-03:39:04
INFO:tensorflow:Saving dict for global step 5000: average_loss = 71.996956, global_step = 5000, label/mean = 22.57451, lo

{'average_loss': 71.996956,
 'global_step': 5000,
 'label/mean': 22.57451,
 'loss': 71.996956,
 'prediction/mean': 22.179829}

# Answer of Q2 is 'average_loss': 71.996956

Q3 Retrain the model of Q2 (from scratch) for 10000 steps instead of 5000 steps. Evaluate the trained model on both the training data and the test data. What is the range of difference between the average loss on the training data and the test data

In [14]:
config = tf.estimator.RunConfig(tf_random_seed=1)

model2=tf.estimator.DNNRegressor(
    feature_columns=feature_columns,
    hidden_units=[15],
    config=config
)

INFO:tensorflow:Using config: {'_model_dir': '/tmp/tmp4f0qv7lm', '_tf_random_seed': 1, '_save_summary_steps': 100, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_session_config': allow_soft_placement: true
graph_options {
  rewrite_options {
    meta_optimizer_iterations: ONE
  }
}
, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_train_distribute': None, '_device_fn': None, '_protocol': None, '_eval_distribute': None, '_experimental_distribute': None, '_experimental_max_worker_delay_secs': None, '_service': None, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7f2df002de80>, '_task_type': 'worker', '_task_id': 0, '_global_id_in_cluster': 0, '_master': '', '_evaluation_master': '', '_is_chief': True, '_num_ps_replicas': 0, '_num_worker_replicas': 1}


In [15]:
model2.train(
input_fn=lambda:input_fn(train,train_y,training=True),
    steps=10000
)

INFO:tensorflow:Calling model_fn.

If you intended to run this layer in float32, you can safely ignore this warning. If in doubt, this warning is likely only an issue if you are porting a TensorFlow 1.X model to TensorFlow 2.

To change all layers to have dtype float64 by default, call `tf.keras.backend.set_floatx('float64')`. To change just this layer, pass dtype='float64' to the layer constructor. If you are the author of this layer, you can disable autocasting by passing autocast=False to the base Layer constructor.

INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 0 into /tmp/tmp4f0qv7lm/model.ckpt.
INFO:tensorflow:loss = 1227.7604, step = 0
INFO:tensorflow:global_step/sec: 234.666
INFO:tensorflow:loss = 192.87956, step = 100 (0.428 sec)
INFO:tensorflow:global_step/sec: 258.87
INFO:tensorflow

In [16]:
model2.evaluate(
input_fn=lambda:input_fn(test,test_y,training=False))

INFO:tensorflow:Calling model_fn.

If you intended to run this layer in float32, you can safely ignore this warning. If in doubt, this warning is likely only an issue if you are porting a TensorFlow 1.X model to TensorFlow 2.

To change all layers to have dtype float64 by default, call `tf.keras.backend.set_floatx('float64')`. To change just this layer, pass dtype='float64' to the layer constructor. If you are the author of this layer, you can disable autocasting by passing autocast=False to the base Layer constructor.

INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2020-10-27T03:46:28Z
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from /tmp/tmp4f0qv7lm/model.ckpt-10000
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Finished evaluation at 2020-10-27-03:46:28
INFO:tensorflow:Saving dict for global step 10000: average_loss = 63.046173, global_step = 10000, label/mean = 22.57451,

{'average_loss': 63.046173,
 'global_step': 10000,
 'label/mean': 22.57451,
 'loss': 63.046173,
 'prediction/mean': 22.33523}

# **Take difference of both losses and see**
**Less than 25**